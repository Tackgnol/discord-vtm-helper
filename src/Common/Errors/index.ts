export * from './FirebaseError';
export * from './InvalidInputError';
