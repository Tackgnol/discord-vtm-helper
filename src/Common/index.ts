export * from './adminMessages';
export * from './isAdminCommand';
export * from './parseCommandMessage';
export * from './parseEventMessage';
export * from './Play';
export * from './reactionNumbers';
export * from './RichEmbeds';
