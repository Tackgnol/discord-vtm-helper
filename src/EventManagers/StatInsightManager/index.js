const { isNil, isEmpty } = require('lodash');
const playerStats = require('../../Resources/playerToStat.json');

class StatInsightManager {
	constructor(message, channel) {
		this.channel = channel;
		this.message = message;
	}

	checkStat(statName, minValue, successMessage) {
		const messageChanel = this.channel;
		if (messageChanel.type === 'text') {
			const channelMembers = messageChanel.members;
			if (!isEmpty(channelMembers)) {
				channelMembers.forEach(p => {
					const thisPlayer = playerStats[p.user.username];
					if (!isNil(thisPlayer) && thisPlayer[statName] >= minValue) {
						p.send(`(${statName}:${thisPlayer[statName]}) ${successMessage}`);
					}
				});
			}
		}
		return;
	}
}

module.exports = StatInsightManager;
