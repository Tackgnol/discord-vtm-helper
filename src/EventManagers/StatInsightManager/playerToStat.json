{
  "Tackgnol": {
    "Politics": 4,
    "Occult": 1
  }
}
