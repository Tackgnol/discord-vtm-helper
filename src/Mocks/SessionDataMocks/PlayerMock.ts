import { Player } from '../../Models/GameData';
import { testNPCWithFacts, testNPCWithoutFacts } from './NPCMocks';

export const playerMock: Player = {
	npcSet: [testNPCWithFacts, testNPCWithoutFacts],
	discordUserName: 'test',
	statisticsSet: [{ name: 'test', value: 2 }],
	id: '0',
};
