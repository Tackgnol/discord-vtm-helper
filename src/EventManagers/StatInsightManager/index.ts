import { find, isEmpty, isNil } from 'lodash';
import { statInsightRichEmbed } from '../../Common/RichEmbeds';
import { Message, TextChannel } from 'discord.js';

class StatInsightManager {
	constructor(private channel: TextChannel, private gameId: string) {
		this.channel = channel;
		this.gameId = gameId;
	}

	async checkStat(statName: string, minValue: number, successMessage: string) {
		const channelMembers = this.channel.members;
		if (!isEmpty(channelMembers)) {
			for (const [snowflake, member] of channelMembers) {
				const thisPlayer = await global.service.GetPlayer(snowflake, this.gameId);
				const statValue = thisPlayer && find(thisPlayer.statisticsSet, s => s.name === statName);
				if (!isNil(statValue) && statValue.value >= minValue) {
					member.send(statInsightRichEmbed(statName, minValue, successMessage));
				}
			}
		}
	}
}

export default StatInsightManager;
