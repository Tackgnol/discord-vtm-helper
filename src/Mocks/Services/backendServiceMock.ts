import { IService } from '../../Services/IService';
import { IGlobalTest, INarration, INPC, IPlayer, IStat, IStatInsight, IVersionOption } from '../../Models/GameData';
import { IGame, ISessionData } from '../../Models/AppModels';
import { testNPCWithFacts, testNPCWithoutFacts } from '../SessionDataMocks/NPCMocks';
import { globalTestWithShortCircuitMock } from '../SessionDataMocks/GlobalTestMocks';
import { basicNarration } from '../SessionDataMocks/NarrationMocks';
import { playerMock } from '../SessionDataMocks/PlayerMock';
import { statInsightMock } from '../SessionDataMocks/StatInsightMocks';
import { activeSessionMock } from '../SessionDataMocks/ActiveSessionMock';
import { gameMock } from '../GameMock';

export class BackendServiceMock implements IService {
	AddFactsToNPC(playerId: string, npc: string, facts: string[], gameId: string): Promise<INPC> {
		if (!(playerId || npc || facts || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve({ ...testNPCWithFacts, facts: [...testNPCWithFacts.facts, ...facts] });
	}

	AddGlobalTest(
		name: string,
		message: string,
		shortCircuit: boolean,
		replyPrefix: string,
		globaltestoptionSet: IVersionOption[],
		channelId: string,
		gameId: string
	): Promise<IGlobalTest> {
		if (!(name || message || shortCircuit || replyPrefix || globaltestoptionSet || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(globalTestWithShortCircuitMock);
	}

	AddNPC(name: string, callName: string, image: string, description: string, gameId: string): Promise<Omit<INPC, 'facts'>> {
		if (!(name || callName || image || description || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(testNPCWithoutFacts);
	}

	AddNarration(name: string, image: string, narrationText: string, channelId: string, gameId: string): Promise<INarration> {
		if (!(name || image || narrationText || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(basicNarration);
	}

	AddPlayer(name: string, id: string, statArray: IStat[], gameId: string): Promise<IPlayer> {
		if (!(name || id || statArray || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	AddStatInsight(
		name: string,
		stat: string,
		value: number,
		message: string,
		channelId: string,
		gameId: string
	): Promise<IStatInsight> {
		if (!(name || stat || value || message || channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(statInsightMock);
	}

	AssignGameAdmin(playerId: string, channelId: string, gameId: string): Promise<ISessionData> {
		if (!(playerId || channelId || gameId)) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetEvents(channelId: string, gameId: string): Promise<ISessionData> {
		if (!(channelId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(activeSessionMock);
	}

	GetPlayer(playerId: string, gameId: string): Promise<IPlayer> {
		if (!(playerId || gameId)) {
			return Promise.reject('Insufficient data passed');
		}
		return Promise.resolve(playerMock);
	}

	GetUserChannels(userId: string): Promise<IGame[]> {
		if (!userId) {
			Promise.reject('Insufficient data passed');
		}
		return Promise.resolve([gameMock]);
	}
}
