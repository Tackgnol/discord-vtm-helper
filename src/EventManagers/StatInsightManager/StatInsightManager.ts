import { find, isEmpty, isNil } from 'lodash';
import { statInsightRichEmbed } from '../../Common';
import { MessageList, ReplyType } from '../../Models/AppModels';
import { InvalidInputError } from '../../Common/Errors/InvalidInputError';
import { IService } from '../../Services/IService';

class StatInsightManager {
	constructor(private service: IService) {}
	async checkStat(statName: string, minValue: number, successMessage: string, gameId: string, channelMembers: string[]) {
		const messageList: MessageList[] = [];
		if (!isEmpty(channelMembers)) {
			for (const member of channelMembers) {
				const thisPlayer = await this.service.GetPlayer(member, gameId);
				const statValue = thisPlayer && find(thisPlayer.statisticsSet, s => s.name === statName);
				if (!isNil(statValue) && statValue.value >= minValue) {
					messageList.push({ recipient: member, message: statInsightRichEmbed(statName, minValue, successMessage) });
				}
			}
			return { type: ReplyType.Multi, value: messageList };
		} else {
			throw new InvalidInputError('Channel has no members!');
		}
	}
}

export default StatInsightManager;
