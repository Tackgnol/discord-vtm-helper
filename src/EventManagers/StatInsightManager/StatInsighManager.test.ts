import { BackendServiceMock } from '../../Mocks/Services/backendServiceMock';
import StatInsightManager from './StatInsightManager';
import { statInsightMockHigh, statInsightMockLow } from '../../Mocks/SessionDataMocks/StatInsightMocks';
import { playerMock } from '../../Mocks/SessionDataMocks/PlayerMock';
import { expect } from 'chai';
import { InvalidInputError } from '../../Common/Errors';

describe('EventManagers >> StatInsightManager >> StatInsightManager', () => {
	const service = new BackendServiceMock();
	const manager = new StatInsightManager(service);
	it('Returns an array of passing players if there are passing players', async () => {
		const { statName, successMessage, minValue } = statInsightMockLow;
		const result = await manager.checkStat(statName, minValue, successMessage, 'test', [playerMock.id]);
		expect(result.value.length).to.not.eq(0);
		expect(result).toMatchSnapshot();
	});
	it('Returns an empty array if no players are passing', async () => {
		const { statName, successMessage, minValue } = statInsightMockHigh;
		const result = await manager.checkStat(statName, minValue, successMessage, 'test', [playerMock.id]);
		expect(result.value.length).to.eq(0);
		expect(result).toMatchSnapshot();
	});
	it('Throws an error if there are no players to test', () => {
		const { statName, successMessage, minValue } = statInsightMockLow;
		expect(manager.checkStat(statName, minValue, successMessage, 'test', [])).to.be.rejectedWith(InvalidInputError);
	});
});
