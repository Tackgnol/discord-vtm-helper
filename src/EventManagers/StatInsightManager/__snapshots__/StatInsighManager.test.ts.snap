// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventManagers >> StatInsightManager >> StatInsightManager Returns an array of passing players if there are passing players 1`] = `
Object {
  "type": 2,
  "value": Array [
    Object {
      "message": Object {
        "author": null,
        "color": 9044739,
        "description": "winner is you!",
        "fields": Array [
          Object {
            "inline": false,
            "name": "test",
            "value": "1",
          },
        ],
        "footer": null,
        "image": null,
        "thumbnail": null,
        "timestamp": null,
        "title": "All your hard work has payed of... Use your insight",
        "type": "rich",
        "url": null,
      },
      "recipient": "0",
    },
  ],
}
`;

exports[`EventManagers >> StatInsightManager >> StatInsightManager Returns an empty array if no players are passing 1`] = `
Object {
  "type": 2,
  "value": Array [],
}
`;
